File: frontend/src/types/stock.ts
export interface PriceBar {
  date: string;
  open: number | null;
  high: number | null;
  low: number | null;
  close: number | null;
  volume: number | null;
}

export interface IndicatorPoint {
  date: string;
  value: number | null;
}

export interface MACDPoint {
  date: string;
  macd: number | null;
  signal: number | null;
  histogram: number | null;
}

export interface StockSummary {
  ticker: string;
  name: string;
  sector: string;
  price: number | null;
  change_pct: number | null;
  market_cap: number | null;
  pe_ratio: number | null;
  pb_ratio: number | null;
  roe: number | null;
  revenue_growth: number | null;
  dividend_yield: number | null;
  volume: number | null;
  rsi: number | null;
}

export interface StockDetail extends StockSummary {
  macd: number | null;
  macd_signal: number | null;
  macd_histogram: number | null;
  sma20: number | null;
  sma50: number | null;
  sma200: number | null;
  ema12: number | null;
  ema26: number | null;
  volume_sma20: number | null;
  description: string | null;
  website: string | null;
  employees: number | null;
  revenue: number | null;
  net_income: number | null;
  total_assets: number | null;
  total_debt: number | null;
  price_history: PriceBar[];
  rsi_history: IndicatorPoint[];
  macd_history: MACDPoint[];
}

File: backend/models/stock.py
from typing import Optional
from pydantic import BaseModel


class PriceBar(BaseModel):
    date: str
    open: Optional[float] = None
    high: Optional[float] = None
    low: Optional[float] = None
    close: Optional[float] = None
    volume: Optional[float] = None


class IndicatorPoint(BaseModel):
    date: str
    value: Optional[float] = None


class MACDPoint(BaseModel):
    date: str
    macd: Optional[float] = None
    signal: Optional[float] = None
    histogram: Optional[float] = None


class StockSummary(BaseModel):
    ticker: str
    name: str
    sector: str
    price: Optional[float] = None
    change_pct: Optional[float] = None
    market_cap: Optional[float] = None
    pe_ratio: Optional[float] = None
    pb_ratio: Optional[float] = None
    roe: Optional[float] = None
    revenue_growth: Optional[float] = None
    dividend_yield: Optional[float] = None
    volume: Optional[float] = None
    rsi: Optional[float] = None


class StockDetail(StockSummary):
    macd: Optional[float] = None
    macd_signal: Optional[float] = None
    macd_histogram: Optional[float] = None
    sma20: Optional[float] = None
    sma50: Optional[float] = None
    sma200: Optional[float] = None
    ema12: Optional[float] = None
    ema26: Optional[float] = None
    volume_sma20: Optional[float] = None
    description: Optional[str] = None
    website: Optional[str] = None
    employees: Optional[int] = None
    revenue: Optional[float] = None
    net_income: Optional[float] = None
    total_assets: Optional[float] = None
    total_debt: Optional[float] = None
    price_history: list[PriceBar] = []
    rsi_history: list[IndicatorPoint] = []
    macd_history: list[MACDPoint] = []
